### PR TITLE
feat(site-builder): add preproc to publish directories with files

### LINKED ISSUE #84: Add a pre-processor to quickly build "filebrowser" Walrus Sites
The idea would be to point this to a directory, this generates an `index.html` file for each directory and links between them, then generates a Walrus Site from that. Would be ideal for hosting and updating files; e.g., for something like https://github.com/MystenLabs/walrus-docs/issues/64.

### PR DESCRIPTION
**Pull request by giac-mysten**
Mimics Python's `list_directory` function when calling `http.server` on a directory without `index.html`.

Closes #84

### CODE REVIEW

**Comment by giac-mysten** [comment]
Result site after running `publish` with the `--preprocess` flag:

https://68u1x1ff5rp69d2kgxxmpg9nh0030r31eojs546ykq59nulvv8.walrus.site/

File: site-builder/src/preprocessor.rs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        for c in contents {
+            body.push_str(&format!("<li>{}</li>\n", c));
+        }
+        body.push_str("</ul>\n");
+        body.push_str("<hr>\n");
+
+        Ok(format!(
+            "<!DOCTYPE html>\n<html>\n<head>\n{}\n</head>\n<body>\n{}\n{}</body>\n</html>",
+            title, h1, body
+        ))
```
Comment (mlegner) [inline site-builder/src/preprocessor.rs:112]: Nit: How about this shorter version?
```suggestion
        let title_string = format!(
            "Directory listing for {}",
            full_path_to_resource_path(&self.path, root)?
        );

        let mut contents: Vec<String> = self
            .contents
            .iter()
            .map(|p| Ok(format!("<li>{}</li>", Self::path_to_html(p, root)?)))
            .collect::<Result<_>>()?;
        contents.sort();
        let body = contents.join("\n");

        Ok(format!(
            "<!DOCTYPE html>\n<html>\n<head>\n<title>{title_string}</title>\n</head>\n\
            <body>\n<h1>{title_string}</h1>\n<hr>\n<ul>\n{body}\n</ul>\n<hr>\n</body>\n</html>",
        ))
```

File: site-builder/src/main.rs [hunk 2]
```diff
@@ -158,6 +171,11 @@ enum Commands {
     },
     /// Show the pages composing the site at the given object ID.
     Sitemap { object: ObjectID },
+    /// Preprocess the directory, creating and linking index files.
+    ///
+    /// This command allows to publish directories as sites. Warning: Rewrites all `index.html`
+    /// files.
+    Preprocess { path: PathBuf },
```
Comment (mlegner) [inline site-builder/src/main.rs:178]: Suggestion: I think we could be a bit more specific than "Preprocess", like "ListDirectory" or something like this. Also, the warning here and for the options above is only visible with "--help" but not with "-h".

**Review by mlegner — COMMENTED** [review]
Amazing, thanks a lot for adding this. A few suggestions for minor improvements.

File: site-builder/src/main.rs [hunk 3]
```diff
@@ -127,6 +130,11 @@ enum Commands {
...
         #[clap(long, default_value_t = 1)]
         epochs: u64,
+        /// Preprocess the directory before publishing.
+        ///
+        /// See the `preprocess` command. Warning: Rewrites all `index.html` files.
+        #[clap(long, action)]
+        preprocess: bool,
     },
     /// Update an existing site
     Update {
```
[the remark this answers is not in the document]
Comment (giac-mysten) [reply site-builder/src/main.rs:140]: Yep, I agree. I had made a note of this. Opened #108

File: site-builder/src/preprocessor.rs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+use crate::site::resource::full_path_to_resource_path;
+
+pub struct Preprocessor;
+
+impl Preprocessor {
+    pub fn iter_dir(path: &Path) -> Result<Vec<DirNode>> {
+        let mut nodes = vec![];
+        let items = std::fs::read_dir(path)?;
+        let mut cur_node = DirNode::new(path.to_path_buf());
+        for item in items.flatten() {
```
[the remark this answers is not in the document]
Comment (giac-mysten) [reply site-builder/src/preprocessor.rs:21]: Why a debug assert only? I can also return an error? Although `read_dir` already returns an error if `!path.is_dir()`

File: site-builder/src/preprocessor.rs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    fn index_path(&self) -> PathBuf {
+        self.path.join("index.html")
+    }
+
+    fn path_to_html(path: &Path, root: &Path) -> Result<String> {
+        let mut link_name = path
+            .file_name()
+            .ok_or(anyhow::anyhow!("no file name"))?
```
[the remark this answers is not in the document]
Comment (giac-mysten) [reply site-builder/src/preprocessor.rs:71]: Indeed, this should only be triggered if the path is root `/` or ends in `..`, eg `/some/path/..` which should never happen. Improved the error a bit.

File: site-builder/src/main.rs (hunk 2 above)
Comment (giac-mysten) [reply site-builder/src/main.rs]: good points, changed.

File: site-builder/src/preprocessor.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (giac-mysten) [reply site-builder/src/preprocessor.rs:22]: Added, regardless :)

[the remark this answers is not in the document]
Comment (mlegner) [reply site-builder/src/preprocessor.rs:22]: I just wanted to clarify the expectation. But I don't really care much about the exact form in this case.

**Review by mlegner — APPROVED** [review]
Thanks a lot for the changes, @giac-mysten!

I will use this to address https://github.com/MystenLabs/walrus-docs/issues/64. 🙂

**Pull request merged**